Move dealiasing from Uncurry to Erasure

retronym opened the pull request:
The motivation is to make the uncurry info tranform an identity
for more method signatures.

A few spots in erasure needed to be updated to account for
the possibility of aliases, by callng normalize and/or using
baseType rather than just assuming that the type args and class
type params line up.

Fixes scala/scala-dev#297

retronym committed:
Move dealiasing from Uncurry to Erasure

The motivation is to make the uncurry info tranform an identity
for more method signatures.

A few spots in erasure needed to be updated to account for
the possibility of aliases, by callng normalize and/or using
baseType rather than just assuming that the type args and class
type params line up.

Fixes scala/scala-dev#297

retronym commented:
```
[error] Uncaught exception when running partest: java.lang.NoSuchMethodError: scala.tools.cmd.Meta$StdOpts$Bash$.<init>(Lscala/tools/cmd/Meta$StdOpts;)V
sbt.ForkMain$ForkError: java.lang.NoSuchMethodError: scala.tools.cmd.Meta$StdOpts$Bash$.<init>(Lscala/tools/cmd/Meta$StdOpts;)V
	at scala.tools.partest.nest.RunnerSpec$.Bash$lzycompute$1(RunnerSpec.scala:50)
	at scala.tools.partest.nest.RunnerSpec$.Bash(RunnerSpec.scala:50)
	at scala.tools.cmd.Meta$StdOpts.$init$(Meta.scala:25)
	at scala.tools.partest.nest.RunnerSpec$.<init>(RunnerSpec.scala:50)
```

adriaanm commented:
```
partest --update-check \
  test/files/run/t6308.scala \
  test/files/run/t6555.scala \
  test/files/neg/t3653.scala \
```

progressions, seems like

retronym commented on src/reflect/scala/reflect/internal/transform/Erasure.scala:
```diff
@@ -111,7 +111,7 @@ trait Erasure {
     protected def eraseDerivedValueClassRef(tref: TypeRef): Type = erasedValueClassArg(tref)
 
-    def apply(tp: Type): Type = tp match {
+    def apply(tp: Type): Type = tp.dealias match {
```
Probably should do `(if (tp.isRefinement) tp.normalize else tp.dealias)) match` upfront here, rather than the subsequent calls to `normalize` below.

retronym replied:
Or maybe not, `RefinedTypes` are not a common case, so it might be better to keep this code direct.

szeiger commented:
@retronym Is this still on track for M2 or should we reschedule it to M3?

retronym commented:
@szeiger I'll move it. This isn't really user visible, it is just an internal cleanup that should marginally help performance.

retronym commented:
Rebased, and re-focused this PR on the subject of dealiasing. The patch to change how uncurry/erasure treats the `IMPLICIT` flag is noted in https://github.com/scala/bug/issues/9848

retronym committed:
Refactor dealiasing in ErasureMap for performance

Because `ErasureMap` already drops RewrappingTypeProxy-s for their
underlying types, the only `dealiasing` that we need to call is
for a type ref to an alias symbol.

retronym commented:
Comparative benchmarks scheduled to [appear soon](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?var-scalaSha=25b9e2b7d3%7Cccbacdd0b3%7C6d297645f8&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench@scalabench@&from=1500851507539&to=1503270707539).

retronym commented:
The performance gains are modest but measurable.

adriaanm commented on src/reflect/scala/reflect/internal/transform/Erasure.scala:
```diff
@@ -142,8 +146,8 @@ trait Erasure {
-      case RefinedType(parents, decls) =>
-        apply(mergeParents(parents))
+      case rt : RefinedType =>
+        apply(mergeParents(rt.normalize.parents)) // RefinedType#normalize
```
Would it make sense to move RefinedType's `normalizeImpl` here? Also, could we do less work when doing Java-style erasure?

retronym replied:
`JavaErasureMap` overrides `mergeParents`:

```scala
    /** In java, always take the first parent.
     *  An intersection such as `Object with Trait` erases to Object.
     */
    def mergeParents(parents: List[Type]): Type =
      if (parents.isEmpty) ObjectTpe
      else parents.head
```

Rather than:

```scala
    /** In scala, calculate a useful parent.
     *  An intersection such as `Object with Trait` erases to Trait.
     */
    def mergeParents(parents: List[Type]): Type =
      intersectionDominator(parents)
```

The erasure type map is run (just) prior to `phase.erasedTypes`, so `RefinedType.normalize` leans on an internal cache:

```scala
    final override def normalize: Type =
      if (phase.erasedTypes) normalizeImpl
      else {
        if (normalized eq null) normalized = normalizeImpl
        normalized
      }
```

If this cache were null at the time we called `normalize`. If `normalize` is not an identity (ie we need to flatten nested refinements), there is a small inefficiency in:
  - computing the flat list of normalized parents and creating a new `RefinedType` with them and the `decls`
  - ... only to extract the `parents` and discard the rest in `ErasureMap`
  - ... only to use the first parent in `JavaErasureMap`.

I believe most of the time `RefinedType.normalize` is an identity, so we're unlikely to see a performance win from this.

adriaanm reviewed (approved):
LGTM, one question that could also be addressed later.

retronym commented:
Let me `jardiff` this before we merge.

retronym commented:
`NeedsSigCollector` is not aware of alias types, so we're missing some generic signatures after this patch.

retronym committed:
Add a test case for generic signatures of alias-typed members

retronym commented:
Even after my latest change, we're missing generic signatures on static forwarders for module methods that with refinement + typealias.

```
--- a/scala/util/control/Exception.class.asm
+++ b/scala/util/control/Exception.class.asm
@@ -123,8 +123,6 @@
     MAXLOCALS = 1

   // access flags 0x9
-  // signature <Ex:Ljava/lang/Throwable;T:Ljava/lang/Object;>(Lscala/Function1<TEx;Ljava/lang/Object;>;Lscala/Function1<TEx;TT;>;Lscala/reflect/ClassTag<TEx;>;)Lscala/PartialFunction<Ljava/lang/Throwable;TT;>;
-  // declaration: scala.PartialFunction<java.lang.Throwable, T> mkCatcher<Ex extends java.lang.Throwable, T>(scala.Function1<Ex, java.lang.Object>, scala.Function1<Ex, T>, scala.reflect.ClassTag<Ex>)
   public static mkCatcher(Lscala/Function1;Lscala/Function1;Lscala/reflect/ClassTag;)Lscala/PartialFunction;
     GETSTATIC scala/util/control/Exception$.MODULE$ : Lscala/util/control/Exception$;
     ALOAD 0
```

minimizes to:

```
object Exception {
  type Catcher[+T] = PartialFunction[Throwable, T]
  def mkCatcher[Ex <: Throwable, T](isDef: Ex => Boolean, f: Ex => T) = new Catcher[T] {
    def isDefinedAt(x: Throwable) = true
    def apply(x: Throwable): T = ???
  }
}
```

The difference is buried in:

```scala
      def staticForwarderGenericSignature: String = {
        // scala/bug#3452 Static forwarder generation uses the same erased signature as the method if forwards to.
        // By rights, it should use the signature as-seen-from the module class, and add suitable
        // primitive and value-class boxing/unboxing.
        // But for now, just like we did in mixin, we just avoid writing a wrong generic signature
        // (one that doesn't erase to the actual signature). See run/t3452b for a test case.
        val memberTpe = enteringErasure(moduleClass.thisType.memberInfo(m))
        val erasedMemberType = erasure.erasure(m).apply(memberTpe)
        if (erasedMemberType =:= m.info)
          getGenericSignature(m, moduleClass, memberTpe)
        else null
      }
```

Which ends up with:

```
erasedMemberType = (isDef: Function1, f: Function1)Object
m.info = (isDef: Function1, f: Function1)PartialFunction
```

Furthermore, specialization is damaged:

```
diff --git a/scala/tools/nsc/ast/parser/Scanners$ParensAnalyzer.class.asm b/scala/tools/nsc/ast/parser/Scanners$ParensAnalyzer.class.asm
index c653301..0db90df 100644
--- a/scala/tools/nsc/ast/parser/Scanners$ParensAnalyzer.class.asm
+++ b/scala/tools/nsc/ast/parser/Scanners$ParensAnalyzer.class.asm
@@ -1732,12 +1732,14 @@
     ISTORE 19
     ALOAD 0
     INVOKEVIRTUAL scala/tools/nsc/ast/parser/Scanners$ParensAnalyzer.nextToken ()V
-    NEW scala/Tuple2$mcII$sp
+    NEW scala/Tuple2
     DUP
     ILOAD 19
+    INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
     ALOAD 4
     GETFIELD scala/runtime/IntRef.elem : I
-    INVOKESPECIAL scala/Tuple2$mcII$sp.<init> (II)V
+    INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
+    INVOKESPECIAL scala/Tuple2.<init> (Ljava/lang/Object;Ljava/lang/Object;)V
     GOTO L22
```

I'm going to close this PR on account of the rapidly increasing worm population in the can I'm attempting to open.

retronym commented:
If we try to pick this up again, I think we should leave refined type normalization (which including dealiasing of its parents) in the `uncurry` info transform. We'll still need to adapt specialization (perhaps it is just the `specializedType` type map that needs alias awareness.
